src/shared/common/CalendarHeatmap.js

```diff
@@ -42,13 +42,80 @@ const styles = StyleSheet.create({
 
 const daysOfWeek = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
 
-function dateToStringKey(date: Date) {
+function getDateSafe(value: string | number | Date) {
+    if (typeof value === 'string') {
+        return new Date(`${value}T00:00`);
+    }
+
+    return new Date(value);
+}
+
+function resolveTime(
+    date: string | number | Date,
+    resolution: 'day' | 'month' | 'year',
+) {
+    const newDate = getDateSafe(date);
+
+    if (
+        resolution === 'day' ||
+        resolution === 'month' ||
+        resolution === 'year'
+    ) {
+        newDate.setHours(0);
+        newDate.setMinutes(0);
+        newDate.setSeconds(0);
+        newDate.setMilliseconds(0);
+    }
+    if (resolution === 'month' || resolution === 'year') {
+        newDate.setDate(1);
+    }
+    if (resolution === 'year') {
+        newDate.setMonth(0);
+    }
+    return newDate;
+}
+
+function getTimestamps(
+    startDate: string | number | Date,
+    endDate: string | number | Date,
+    resolution: 'day' | 'month' | 'year',
+) {
+    const sanitizedStartDate = resolveTime(startDate, resolution);
+    const sanitizedEndDate = resolveTime(endDate, resolution);
+
+    const timestamps = [sanitizedStartDate.getTime()];
+
+    let increment = 1;
+    while (true) {
```

Comment:
Any risk of an infinite loop here? I'm thinking if the start and end dates are very far apart, or if `endDate < startDate` then you'd never exit and the app would hang. Is it worth adding some checks to prevent that?

Reply:
@laurentS since startDate and endDate are always 35 days apart, there isn't any chance of infinite loop here.

Reply:
> @laurentS since startDate and endDate are always 35 days apart, there isn't any chance of infinite loop here.

They always are, until they are not. Some other dev might call this function without realising this expectation (which isn't documented), and now you have a bug :)

Reply:
@laurentS we have removed the possibility of an infinite loop here.